File: rust/src/392.rs
//392. Is Subsequence

impl Solution {
    pub fn is_subsequence(s: String, t: String) -> bool {
        let (mut i, mut j) = (0, 0);
        let s_chars: Vec<char> = s.chars().collect();
        let t_chars: Vec<char> = t.chars().collect();

        while i < s_chars.len() && j < t_chars.len() {
            if s_chars[i] == t_chars[j] {
                i += 1;
            }
            j += 1;
        }

        i == s_chars.len()
    }
}
//sub: https://leetcode.com/problems/is-subsequence/submissions/1594060955/?envType=study-plan-v2&envId=top-interview-150
//sol: https://leetcode.com/problems/is-subsequence/solutions/6606224/392-is-subsequence-by-nalindalal2004-egtt/

File: ts/src/392.ts
//2300. Successful Pairs of Spells and Potions
function successfulPairs(
  spells: number[],
  potions: number[],
  success: number,
): number[] {
  potions.sort((a, b) => a - b);
  const m = potions.length;
  const res: number[] = [];

  for (const spell of spells) {
    const minPotion = Math.ceil(success / spell);

    // Binary search for the first potion >= minPotion
    let left = 0,
      right = m - 1,
      idx = m;
    while (left <= right) {
      const mid = Math.floor((left + right) / 2);
      if (potions[mid] >= minPotion) {
        idx = mid;
        right = mid - 1;
      } else {
        left = mid + 1;
      }
    }

    res.push(m - idx); // number of successful potions
  }

  return res;
}

//sub: https://leetcode.com/problems/successful-pairs-of-spells-and-potions/submissions/1794950306/?envType=daily-question&envId=2025-10-08
//sol: https://leetcode.com/problems/successful-pairs-of-spells-and-potions/solutions/7258068/2300-successful-pairs-of-spells-and-poti-uewb/
